server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java

```diff
@@ -0,0 +1,165 @@
+        return this.positionIncrementGap;
+    }
+
+    @Override
+    public int getOffsetGap(String field) {
+        if (this.offsetGap < 0) {
+            return super.getOffsetGap(field);
+        }
+        return this.offsetGap;
+    }
+
+    public AnalysisMode getAnalysisMode() {
+        return this.components.analysisMode();
+    }
+
+    @Override
+    protected Reader initReaderForNormalization(String fieldName, Reader reader) {
+        final AnalyzerComponents components = getComponents();
+        for (CharFilterFactory charFilter : components.getCharFilters()) {
+            reader = charFilter.normalize(reader);
+        }
+        return reader;
+    }
+
+    @Override
+    protected TokenStream normalize(String fieldName, TokenStream in) {
+        final AnalyzerComponents components = getComponents();
+        TokenStream result = in;
+        for (TokenFilterFactory filter : components.getTokenFilters()) {
+            result = filter.normalize(result);
+        }
+        return result;
+    }
+
+    private boolean shouldReload(AnalyzerComponents source) {
+        return this.components != source;
+    }
+
+    public synchronized void reload(String name,
+                                    Settings settings,
+                                    final Map<String, TokenizerFactory> tokenizers,
+                                    final Map<String, CharFilterFactory> charFilters,
+                                    final Map<String, TokenFilterFactory> tokenFilters) {
+        AnalyzerComponents components = CustomAnalyzerProvider.createComponents(name, settings, tokenizers, charFilters, tokenFilters);
+        this.components = components;
+    }
+
+    @Override
+    public void close() {
+        storedComponents.close();
+    }
+
+    private void setStoredComponents(AnalyzerComponents components) {
+        storedComponents.set(components);
+    }
+
+    private AnalyzerComponents getStoredComponents() {
+        return storedComponents.get();
+    }
+
+    @Override
+    protected TokenStreamComponents createComponents(String fieldName) {
+        final AnalyzerComponents components = getStoredComponents();
+        Tokenizer tokenizer = components.getTokenizerFactory().create();
+        TokenStream tokenStream = tokenizer;
+        for (TokenFilterFactory tokenFilter : components.getTokenFilters()) {
+            tokenStream = tokenFilter.create(tokenStream);
+        }
+        return new TokenStreamComponents(tokenizer, tokenStream);
+    }
+
+    @Override
+    protected Reader initReader(String fieldName, Reader reader) {
+        final AnalyzerComponents components = getStoredComponents();
+        if (components.getCharFilters() != null && components.getCharFilters().length > 0) {
+            for (CharFilterFactory charFilter : components.getCharFilters()) {
+                reader = charFilter.create(reader);
+            }
+        }
+        return reader;
+    }
+}
```

Comment:
Given how much of this is shared with CustomAnalyzer, I wonder if we ought to make CustomAnalyzer take AnalyzerComponents as well and then have ReloadableCustomAnalyzer extend it?

Reply:
That was my first solution in https://github.com/elastic/elasticsearch/pull/40782 which Jim was somewhat against. One reason was that this changed CustomAnalyzer to non-final, also it changed the CustomAnalyzers API a bit which we said we'd try to avoid. Happy to go the other way again but would like to get some resolution before reworking everything again here.

Reply:
I am not worried about the sharing of code, the way we use the components in the reloadable analyzer is quite different and would be difficult to generalize without changing the custom analyzer. However I wonder if we could have a base class similar to `TokenFilterComposite` but that would give access to the `AnalyzerComponents` entirely:
````
public class AbstractCustomAnalyzer {
    public abstract AnalyzerComponents[] components();
}
````
This would also resolve my previous comment regarding the way we expose individual component of the `AnalyzerComponents` in the reloadable analyzer.

Reply:
I addressed this in https://github.com/elastic/elasticsearch/pull/42888/commits/eae870fbaf3dc6359f84fe0f63bc3cedf8555142, but decided to go with an interface instead of an abstract class